File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
    int left = 0, maxLen = 0;
    Set<Character> set = new HashSet<>();

    for (int right = 0; right < s.length(); right++) {
        while (set.contains(s.charAt(right))) {
            set.remove(s.charAt(left));
            left++; // shrink window
        }
        set.add(s.charAt(right));
        maxLen = Math.max(maxLen, right - left + 1);
    }
    return maxLen;
    }
}

File: 0003-longest-substring-without-repeating-characters/0003-longest-substring-without-repeating-characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        seen = set()
        left = 0
        res = 0
        for right in range(len(s)):
            while s[right] in seen:
                seen.remove(s[left])
                left += 1
            seen.add(s[right])
            res = max(res, right - left + 1)
        return res
        